Partitioned heat conduction (G+Smo participant)

Crazy-Rich-Meghan opened the pull request:
Summary: Partitioned heat conduction G+Smo (IsoGeometric Analysis) participant
**NOTE**: G+Smo adapter is a submodule of the G+Smo library, to run the tutorial you need to configure G+Smo main library. 
- [x] Updated README.md file, which contains instructions on how to configure G+Smo library with preCICE adapter.
- [x] Added gismo-executable folder to contain the symbolic link for the G+Smo executable.

This PR is marked as a draft because the documentation for the G+Smo adapter requires updates.

Crazy-Rich-Meghan committed:
G+Smo cpp file; run scripts; clean scripts

Crazy-Rich-Meghan committed:
Merge branch 'precice:develop' into partitioned-heat-conduction

Crazy-Rich-Meghan committed:
Update README.md for G+Smo; rename files

Crazy-Rich-Meghan committed:
Fix README.md

uekerman commented:
Let's first iterate on #603 and then here, OK?

MakisH commented:
@Crazy-Rich-Meghan do you plan to continue this development? It seems to currently be in draft state. We are happy to help get it in a good state!

Crazy-Rich-Meghan committed:
Merge branch 'precice:develop' into partitioned-heat-conduction

Crazy-Rich-Meghan committed:
Direct mesh access version for partitioned heat conduction

Crazy-Rich-Meghan commented:
Hi everyone,

Sorry for the delay! I have updated the partitioned heat conduction example from our side, this version supports the direct mesh access feature. To run the benchmark you need to do the following:
    1. Clone G+Smo 
    2.  `cmake .. -DCMAKE_BUILD_TYPE=Release -DGISMO_OPTIONAL="gsPreCICE"`
    3. `cmake --build build --target partitioned-heat-conduction -j 8`
    4.  Install this example systemwide `make install partitioned-heat-conduction`
    5. Go to the `neumann-gismo` folder run the Neumann participant with the run script.
    6. Go to the `dirichlet-gismo` folder run the Dirichlet participant with the run script.

Let me know if you get stuck somewhere!

MakisH commented:
Hi @Crazy-Rich-Meghan,

thanks for the updates! I am now trying to build on my system, based on the stable branch (Commit [ef4c754785cf4949d3a89b916eed807b74b97a9b](https://github.com/gismo/gismo/commit/ef4c754785cf4949d3a89b916eed807b74b97a9b), October 2).

I do get a building error, though (`g++ (Ubuntu 12.3.0-1ubuntu1~22.04.2) 12.3.0`):

```
[ 97%] Building CXX object optional/gsPreCICE/examples/CMakeFiles/partitioned-heat-conduction.dir/partitioned-heat-conduction.cpp.o
/home/gc/repos/misc/gismo/optional/gsPreCICE/examples/partitioned-heat-conduction.cpp: In function ‘int main(int, char**)’:
/home/gc/repos/misc/gismo/optional/gsPreCICE/examples/partitioned-heat-conduction.cpp:110:17: error: base operand of ‘->’ has non-pointer type ‘gismo::gsBasis<double>::domainIter’ {aka ‘gismo::gsDomainIteratorWrapper<double>’}
  110 |     for (; domIt->good(); domIt->next(), k++ )
      |                 ^~
/home/gc/repos/misc/gismo/optional/gsPreCICE/examples/partitioned-heat-conduction.cpp:110:32: error: base operand of ‘->’ has non-pointer type ‘gismo::gsBasis<double>::domainIter’ {aka ‘gismo::gsDomainIteratorWrapper<double>’}
  110 |     for (; domIt->good(); domIt->next(), k++ )
      |                                ^~
/home/gc/repos/misc/gismo/optional/gsPreCICE/examples/partitioned-heat-conduction.cpp:120:15: error: base operand of ‘->’ has non-pointer type ‘gismo::gsBasis<double>::domainIter’ {aka ‘gismo::gsDomainIteratorWrapper<double>’}
  120 |     for (domIt->reset(); domIt->good(); domIt->next(), k++ )
      |               ^~
/home/gc/repos/misc/gismo/optional/gsPreCICE/examples/partitioned-heat-conduction.cpp:120:31: error: base operand of ‘->’ has non-pointer type ‘gismo::gsBasis<double>::domainIter’ {aka ‘gismo::gsDomainIteratorWrapper<double>’}
  120 |     for (domIt->reset(); domIt->good(); domIt->next(), k++ )
      |                               ^~
/home/gc/repos/misc/gismo/optional/gsPreCICE/examples/partitioned-heat-conduction.cpp:120:46: error: base operand of ‘->’ has non-pointer type ‘gismo::gsBasis<double>::domainIter’ {aka ‘gismo::gsDomainIteratorWrapper<double>’}
  120 |     for (domIt->reset(); domIt->good(); domIt->next(), k++ )
      |                                              ^~
/home/gc/repos/misc/gismo/optional/gsPreCICE/examples/partitioned-heat-conduction.cpp:124:29: error: base operand of ‘->’ has non-pointer type ‘gismo::gsBasis<double>::domainIter’ {aka ‘gismo::gsDomainIteratorWrapper<double>’}
  124 |         QuRule->mapTo( domIt->lowerCorner(), domIt->upperCorner(),
      |                             ^~
/home/gc/repos/misc/gismo/optional/gsPreCICE/examples/partitioned-heat-conduction.cpp:124:51: error: base operand of ‘->’ has non-pointer type ‘gismo::gsBasis<double>::domainIter’ {aka ‘gismo::gsDomainIteratorWrapper<double>’}
  124 |         QuRule->mapTo( domIt->lowerCorner(), domIt->upperCorner(),
      |                                                   ^~
gmake[3]: *** [optional/gsPreCICE/examples/CMakeFiles/partitioned-heat-conduction.dir/build.make:76: optional/gsPreCICE/examples/CMakeFiles/partitioned-heat-conduction.dir/partitioned-heat-conduction.cpp.o] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:1940: optional/gsPreCICE/examples/CMakeFiles/partitioned-heat-conduction.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:1947: optional/gsPreCICE/examples/CMakeFiles/partitioned-heat-conduction.dir/rule] Error 2
gmake: *** [Makefile:907: partitioned-heat-conduction] Error 2
```

By the way, I edited your instructions above to add a number next to the `-j` flag. Without a number, it will use as many threads as parallelizable build targets, which might fill the memory completely and make a system crash.

MakisH commented on partitioned-heat-conduction/dirichlet-gismo/clean.sh:
```diff
@@ -0,0 +1,27 @@
+
+echo "Cleaning unnecessary files..."
+
+# Remove precice-profiling directory
+if [ -d "precice-profiling" ]; then
+    rm -rf precice-profiling
+    echo "Deleted 'precice-profiling' folder."
+fi
+
+# Remove precice-run directory
+if [ -d "../precice-run" ]; then
+    rm -rf ../precice-run
+    echo "Deleted 'precice-run' folder."
+fi
```
These should be handled by `clean_precice_logs()`, called by `clean_gismo()` (see https://github.com/precice/tutorials/blob/develop/tools/cleaning-tools.sh)

MakisH commented on partitioned-heat-conduction/dirichlet-gismo/clean.sh:
```diff
@@ -0,0 +1,27 @@
+fi
+
+
+# Remove files ending with .pvd, .vts, .vtp, .log, and .txt
+for ext in pvd vts vtp log txt; do
+    find . -type f -name "*.$ext" -exec rm -f {} \;
+    echo "Deleted all *.$ext files."
+done
```
This is a bit too intrusive (especially for the `.txt` files). Better do a fine-grained extension of `clean_gismo()`.

Same in the `neumann-gismo` directory.

MakisH commented on partitioned-heat-conduction/README.md:
```diff
@@ -37,6 +37,8 @@ You can either couple a solver with itself or different solvers with each other.
 
 * OpenFOAM. This case uses the custom [heatTransfer](https://github.com/precice/tutorials/blob/master/partitioned-heat-conduction/solver-openfoam/heatTransfer.C) solver (find it in `solver-openfoam` and build with `wmake`). Read more details in the [OpenFOAM adapter](https://precice.org/adapter-openfoam-overview.html).
 
+* G+Smo. Inatsll [G+Smo](https://github.com/gismo/gismo).
```
```suggestion
* G+Smo. Install [G+Smo](https://github.com/gismo/gismo).
```

But we need a bit more details, or better link to the adapter documentation:


```suggestion
* G+Smo. Install the [G+Smo adapter](https://precice.org/adapter-gismo.html).
```

MakisH commented on partitioned-heat-conduction/README.md:
````diff
@@ -73,6 +75,38 @@ If you want to use Nutils or OpenFOAM, use `cd dirichlet/neumann-nutils`, respec
+```
+cd neumann-gismo
+./run.sh
+```
````
Could we move most of this content to the adapter documentation, and refer to that?

The path `./gismo_executable/gismo_executable` is assumed by the `run.sh`, so it should probably be documented here. But we should better call that directory `solver-gismo`.

MakisH commented on partitioned-heat-conduction/README.md:
```diff
@@ -85,6 +119,8 @@ For FEniCS you can visualize the content with paraview by opening the `*.pvd` fi
 
 For Nutils, please use the files `Dirichlet-*.vtk` or `Neumann-*.vtk`. Please note that these files contain the temperature as well as the reference solution.
 
+For G+Smo, please use the file `solution.pvd` in both dirichlet-gismo and neumann-gismo directories. 
```
```suggestion
For G+Smo, please use the generated `solution.pvd` files in the dirichlet-gismo and neumann-gismo directories. 
```

(I thought this was a file in this PR for a moment)

MakisH commented on partitioned-heat-conduction/images/tutorial-partitioned-heat-conduction-gismo.gif:
Not yet referred to from the `README.md`. In principle, both FEniCS and G+Smo should give the same solution. If not, we should add the picture in the `README.md` together with a comment.

Crazy-Rich-Meghan committed:
Merge branch 'precice:develop' into partitioned-heat-conduction

Crazy-Rich-Meghan committed:
Update partitioned-heat-conduction/README.md

Co-authored-by: Gerasimos Chourdakis <gerasimos.chourdakis@ipvs.uni-stuttgart.de>

Crazy-Rich-Meghan committed:
Update partitioned-heat-conduction/README.md

Co-authored-by: Gerasimos Chourdakis <gerasimos.chourdakis@ipvs.uni-stuttgart.de>

Crazy-Rich-Meghan commented:
> Hi @Crazy-Rich-Meghan,
> 
> 
> 
> thanks for the updates! I am now trying to build on my system, based on the stable branch (Commit [ef4c754785cf4949d3a89b916eed807b74b97a9b](https://github.com/gismo/gismo/commit/ef4c754785cf4949d3a89b916eed807b74b97a9b), October 2).
> 
> 
> 
> I do get a building error, though (`g++ (Ubuntu 12.3.0-1ubuntu1~22.04.2) 12.3.0`):
> 
> 
> 
> ```
> 
> [ 97%] Building CXX object optional/gsPreCICE/examples/CMakeFiles/partitioned-heat-conduction.dir/partitioned-heat-conduction.cpp.o
> 
> /home/gc/repos/misc/gismo/optional/gsPreCICE/examples/partitioned-heat-conduction.cpp: In function ‘int main(int, char**)’:
> 
> /home/gc/repos/misc/gismo/optional/gsPreCICE/examples/partitioned-heat-conduction.cpp:110:17: error: base operand of ‘->’ has non-pointer type ‘gismo::gsBasis<double>::domainIter’ {aka ‘gismo::gsDomainIteratorWrapper<double>’}
> 
>   110 |     for (; domIt->good(); domIt->next(), k++ )
> 
>       |                 ^~
> 
> /home/gc/repos/misc/gismo/optional/gsPreCICE/examples/partitioned-heat-conduction.cpp:110:32: error: base operand of ‘->’ has non-pointer type ‘gismo::gsBasis<double>::domainIter’ {aka ‘gismo::gsDomainIteratorWrapper<double>’}
> 
>   110 |     for (; domIt->good(); domIt->next(), k++ )
> 
>       |                                ^~
> 
> /home/gc/repos/misc/gismo/optional/gsPreCICE/examples/partitioned-heat-conduction.cpp:120:15: error: base operand of ‘->’ has non-pointer type ‘gismo::gsBasis<double>::domainIter’ {aka ‘gismo::gsDomainIteratorWrapper<double>’}
> 
>   120 |     for (domIt->reset(); domIt->good(); domIt->next(), k++ )
> 
>       |               ^~
> 
> /home/gc/repos/misc/gismo/optional/gsPreCICE/examples/partitioned-heat-conduction.cpp:120:31: error: base operand of ‘->’ has non-pointer type ‘gismo::gsBasis<double>::domainIter’ {aka ‘gismo::gsDomainIteratorWrapper<double>’}
> 
>   120 |     for (domIt->reset(); domIt->good(); domIt->next(), k++ )
> 
>       |                               ^~
> 
> /home/gc/repos/misc/gismo/optional/gsPreCICE/examples/partitioned-heat-conduction.cpp:120:46: error: base operand of ‘->’ has non-pointer type ‘gismo::gsBasis<double>::domainIter’ {aka ‘gismo::gsDomainIteratorWrapper<double>’}
> 
>   120 |     for (domIt->reset(); domIt->good(); domIt->next(), k++ )
> 
>       |                                              ^~
> 
> /home/gc/repos/misc/gismo/optional/gsPreCICE/examples/partitioned-heat-conduction.cpp:124:29: error: base operand of ‘->’ has non-pointer type ‘gismo::gsBasis<double>::domainIter’ {aka ‘gismo::gsDomainIteratorWrapper<double>’}
> 
>   124 |         QuRule->mapTo( domIt->lowerCorner(), domIt->upperCorner(),
> 
>       |                             ^~
> 
> /home/gc/repos/misc/gismo/optional/gsPreCICE/examples/partitioned-heat-conduction.cpp:124:51: error: base operand of ‘->’ has non-pointer type ‘gismo::gsBasis<double>::domainIter’ {aka ‘gismo::gsDomainIteratorWrapper<double>’}
> 
>   124 |         QuRule->mapTo( domIt->lowerCorner(), domIt->upperCorner(),
> 
>       |                                                   ^~
> 
> gmake[3]: *** [optional/gsPreCICE/examples/CMakeFiles/partitioned-heat-conduction.dir/build.make:76: optional/gsPreCICE/examples/CMakeFiles/partitioned-heat-conduction.dir/partitioned-heat-conduction.cpp.o] Error 1
> 
> gmake[2]: *** [CMakeFiles/Makefile2:1940: optional/gsPreCICE/examples/CMakeFiles/partitioned-heat-conduction.dir/all] Error 2
> 
> gmake[1]: *** [CMakeFiles/Makefile2:1947: optional/gsPreCICE/examples/CMakeFiles/partitioned-heat-conduction.dir/rule] Error 2
> 
> gmake: *** [Makefile:907: partitioned-heat-conduction] Error 2
> 
> ```
> 
> 
> 
> By the way, I edited your instructions above to add a number next to the `-j` flag. Without a number, it will use as many threads as parallelizable build targets, which might fill the memory completely and make a system crash.

Hi @Makis, have you pulled the main Gismo? We couldn't replicate the issue (on both MacOS and Ubuntu).

MakisH commented on partitioned-heat-conduction/README.md:
````diff
@@ -73,6 +75,38 @@ If you want to use Nutils or OpenFOAM, use `cd dirichlet/neumann-nutils`, respec
+```
+ - Configure G+Smo 
+```
+cmake .. -DGISMO_OPTIONAL="<other submodules>;gsPreCICE"
````
Do we need other modules for this tutorial? If not, let's remove. If yes, document.

But as I wrote in another comment, this looks like something to move to the adapter documentation.

MakisH commented on partitioned-heat-conduction/README.md:
````diff
@@ -73,6 +75,38 @@ If you want to use Nutils or OpenFOAM, use `cd dirichlet/neumann-nutils`, respec
+```
+- Build the Example
+```
+make partitioned-heat-conduction -j <number of threads to use>
````
This is still failing for me with:

```
commit 62078b48516c4a80869c56caab83c393f13640e8 (HEAD -> stable, origin/stable, origin/HEAD)
Merge: cac651c3 f2565ee4
Author: Hugo Verhelst <38376601+hverhelst@users.noreply.github.com>
Date:   Wed Dec 17 15:31:48 2025 +0100

    Merge pull request #778 from gismo/gsCahnHilliardAssembler
    
    Adding a black-box assembler for the Cahn-Hilliard equation & Color output
```

Here is the full log: [gismo-building-issue.log](https://github.com/user-attachments/files/24532037/gismo-building-issue.log)

MakisH replied:
This worked with gsPreCICE commit e557b51.

MakisH reviewed (commented):
@Crazy-Rich-Meghan I have not forgotten about this PR, and I see there are still some easy points to close. I added a log file with my building issue.

Crazy-Rich-Meghan commented:
Hi @MakisH,

The `domIt->good()` issue has been fixed in gsPreCICE (commit e557b51). Please update your gsPreCICE to the latest version:

```bash
rm -rf optional/gsPreCICE
cmake .. -DGISMO_OPTIONAL="gsPreCICE"
```

MakisH commented on partitioned-heat-conduction/README.md:
````diff
@@ -73,6 +75,38 @@ If you want to use Nutils or OpenFOAM, use `cd dirichlet/neumann-nutils`, respec
+```
+git clone git@github.com:gismo/gismo.git
+cd gismo
+mkdir build
````
```suggestion
cd gismo
mkdir build
cd build
```

MakisH committed:
Update partitioned-heat-conduction/README.md

MakisH commented:
Thanks, @Crazy-Rich-Meghan, this worked! The `perpendicular-flap-vertex-gismo` does not build for me anymore, but I will have another look.

If you have no objections, I will (later) push a few changes directly to this branch, to make the tutorial more consistent with https://precice.org/tutorials-perpendicular-flap-stress.html

MakisH committed:
Merge branch 'develop' into partitioned-heat-conduction

Crazy-Rich-Meghan committed:
Address review comments for G+Smo tutorial

Crazy-Rich-Meghan commented:
Hi @MakisH,

Thanks for the review! I've addressed your comments:

- Updated `clean.sh` scripts to use the standard `clean_gismo()` function
- Simplified the README.md, linking to the adapter documentation
- Updated executable paths in `run.sh` scripts
- Removed the redundant gismo.gif (same result as FEniCS)

Please let me know if there's anything else that needs to be changed.

MakisH committed:
Merge branch 'develop' into partitioned-heat-conduction

Crazy-Rich-Meghan committed:
Fix style issues in run.sh and README

MakisH committed:
Merge branch 'develop' into partitioned-heat-conduction

MakisH commented:
@Crazy-Rich-Meghan I am trying to run this case again, but it is failing immediately with:

```
Patches: 1, degree: 2
Active options:
Options (12):
* DirichletStrategy    (int) = 11       Method for enforcement of Dirichlet BCs [11..14]
* DirichletValues      (int) = 101      Method for computation of Dirichlet DoF values [100..103]
* SameElement         (bool) = 1        Activates optimization if all quadrature points are located in the same element
* bdA                 (real) = 2        Estimated nonzeros per column of the matrix: bdA*deg + bdB
* bdB                  (int) = 1        Estimated nonzeros per column of the matrix: bdA*deg + bdB
* bdO                 (real) = 0.333    Overhead of sparse mem. allocation: (1+bdO)(bdA*deg + bdB) [0..1]
* flipSide            (bool) = 0        Flip side of interface where integration is performed.
* movingInterface     (bool) = 0        Used in interface assembly when interface is not stationary.
* overInt             (bool) = 0        Apply over-integration on boundary elements or not?
* quA                 (real) = 1        Number of quadrature points: quA*deg + quB; For patchRule: Regularity of the target space
* quB                  (int) = 1        Number of quadrature points: quA*deg + quB; For patchRule: Degree of the target space
* quRule               (int) = 1        Quadrature rule used (1) Gauss-Legendre; (2) Gauss-Lobatto; (3) Patch-Rule

Assert `index >= 0 && index < size()` 
/home/gc/repos/misc/gismo/external/gsEigen/src/Core/DenseCoeffsBase.h, line 163 (operator[])
terminate called after throwing an instance of 'std::logic_error'
  what():  GISMO_ASSERT
./run.sh: line 7: 90127 Aborted                 (core dumped) ./gismo-executable -c ../precice-config.xml --plot -s 0
```

I am using a clean build of the latest `dev` branch (but an earlier commit was also failing) - commit https://github.com/gismo/gismo/commit/a43e43468a3fd56c3a11c37c90256c7277ad4fd2

I have configured it with the modules `gsElasticity;gsKLShell;gsPreCICE;gsStructuralAnalysis`.

Does the same state of G+Smo work for you?

MakisH commented on partitioned-heat-conduction/README.md:
````diff
@@ -75,6 +77,27 @@ If you want to use Nutils or OpenFOAM, use `cd dirichlet/neumann-nutils`, respec
+
+```bash
+cd neumann-gismo
+./run.sh
````
While I can run the `partitioned-heat-conduction-direct` case, this case gives me a configuration error early on. Do I need a different configuration or solver executable?

```
root@dfc17ae89aca:/mnt/host/tutorials/partitioned-heat-conduction/neumann-gismo# ./run.sh 
Started on: Wed, 13 May 2026 13:35:46 +0200
Patches: 1, degree: 2
Active options:
Options (12):
* DirichletStrategy    (int) = 11       Method for enforcement of Dirichlet BCs [11..14]
* DirichletValues      (int) = 101      Method for computation of Dirichlet DoF values [100..103]
* SameElement         (bool) = 1        Activates optimization if all quadrature points are located in the same element
* bdA                 (real) = 2        Estimated nonzeros per column of the matrix: bdA*deg + bdB
* bdB                  (int) = 1        Estimated nonzeros per column of the matrix: bdA*deg + bdB
* bdO                 (real) = 0.333    Overhead of sparse mem. allocation: (1+bdO)(bdA*deg + bdB) [0..1]
* flipSide            (bool) = 0        Flip side of interface where integration is performed.
* movingInterface     (bool) = 0        Used in interface assembly when interface is not stationary.
* overInt             (bool) = 0        Apply over-integration on boundary elements or not?
* quA                 (real) = 1        Number of quadrature points: quA*deg + quB; For patchRule: Regularity of the target space
* quB                  (int) = 1        Number of quadrature points: quA*deg + quB; For patchRule: Degree of the target space
* quRule               (int) = 1        Quadrature rule used (1) Gauss-Legendre; (2) Gauss-Lobatto; (3) Patch-Rule

---[precice]  This is preCICE version 3.4.1
---[precice]  Revision info: no-info [git failed to run]
---[precice]  Build type: Release (without debug log)
---[precice]  Working directory "/mnt/host/tutorials/partitioned-heat-conduction/neumann-gismo"
---[precice]  Configuring preCICE with configuration "../precice-config.xml"
---[precice]  I am participant "Neumann"
GISMO_DEBUG: gsPreCICE.h:178, "setMeshVertices": 
setMeshVertices
---[precice] ERROR:  This participant attempteded to set an access region (via "setMeshAccessRegion") on mesh "Dirichlet-Mesh", but mesh "Dirichlet-Mesh" is either not a received mesh or its api access was not enabled in the configuration. setMeshAccessRegion(...) is only valid for (<receive-mesh name="Dirichlet-Mesh" ... api-access="true"/>).
terminate called after throwing an instance of 'precice::Error'
  what():  This participant attempteded to set an access region (via "setMeshAccessRegion") on mesh "Dirichlet-Mesh", but mesh "Dirichlet-Mesh" is either not a received mesh or its api access was not enabled in the configuration. setMeshAccessRegion(...) is only valid for (<receive-mesh name="Dirichlet-Mesh" ... api-access="true"/>).
```

MakisH committed:
Make run.sh scripts consistent

MakisH reviewed (changes requested):
I tried again in an Ubuntu 24.04 container and the `partitioned-heat-conduction-direct` case works (note to self: container ID `dfc17ae89aca`, name `affectionate_sinoussi`).

However, the `partitioned-heat-conduction` still seems to be calling the direct mesh access API. @Crazy-Rich-Meghan am I missing some option? Am I building the wrong solver?

Crazy-Rich-Meghan committed:
Use api-access instead of deprecated direct-access

Crazy-Rich-Meghan committed:
Document the gsPreCICE submodule dependency for the G+Smo build

MakisH committed:
Extend clean_gismo() to clean .vtp, .vts, .pvd files

MakisH reviewed (approved):
Everything works as expected now. Thanks a lot for the (many) updates, @Crazy-Rich-Meghan!
